### PR TITLE
Overwrite slice flag defaults when set

### LINKED ISSUE #160: cli.StringSliceFlag and default value
Hello,
I define my flag as follow:

```go
flFilter = cli.StringSliceFlag{
        Name:  "filter",
        Value: &cli.StringSlice{"flag1", "flag2", "flag3"},
}
```

If I start my binary without arguments, `c.StringSlice("filter")` returns `["flag1", "flag2", "flag3"]` as expected.
If I start my binary with `--filter flag4`,  `c.StringSlice("filter")` returns `["flag1", "flag2", "flag3", "flag4"]`. I find this behaviour odd, I expected it to erase the default value.
Even worst, if I start my binary with `--filter flag1`,  `c.StringSlice("filter")` returns `["flag1", "flag2", "flag3", "flag1"]`.

Is this the expected behavior ? I know I have 3 default value so I can do `c.StringSlice("filter")[3:]` if set, but still it's not perfect.

### LINKED ISSUE #307: EnvVar shadows command line parameters
When EnvVar value is set on global flags, command line parameters has no effective.

### LINKED ISSUE #314: Context.IsSet returns false even if StringSlice flag is set
On my environment (mac OS X 10.10.5), `Context.IsSet` does not work when using `StringSliceFlag` .

```go
// `cmd` is a "sub_command" object
cmd.Flags = []cli.Flag{
	cli.StringFlag{
		Name:   "string_arg, s",
		Value:  "",
		Usage:  "snip",
	},
	cli.StringSliceFlag{
		Name:  "string_slice_arg, a",
		Value: &cli.StringSlice{},
		Usage: "snip",
	},
}

// ...snip

fmt.Println(c.IsSet("string_arg"))
fmt.Println(c.IsSet("string_slice_arg"))
```

```bash
./go_binary sub_command -s str1 -a str2
true
false
```

### LINKED ISSUE #319: values get added to cli.StringSlice if both flag and environment values are given
I encountered this case when I gave a command flag value while the environment variable was also set for the same. I am not sure it is a bug/feature. but decided to log it.

### PR DESCRIPTION
**Pull request by meatballhat**
Closes #160 
Closes #319 
Closes #314 
Closes #307 

*NOTE*: this includes an API change:

### old way, no longer valid:

``` go
s := &StringSlice{"foo", "bar"}
i := &IntSlice{1, 2}
```

### new way:

``` go
s := NewStringSlice("foo", "bar")
i := NewIntSlice(1, 2)
```

### CODE REVIEW

**Comment by jszwedko** [comment]
:+1: :100: 

I thought about this a bit more and it does feel like it would be disruptive to backwards compatibility (since the only people who would have been bitten by this must already be specifying a value), but I can't think of another way to address this bug, so I would like to:

* Tag the current master
* Create a CHANGELOG

before merging this just to ease the pain a little.

**Comment by meatballhat** [comment]
@jszwedko SGTM!  I'm happy to do the tagging and CHANGELOG bits.

**Comment by meatballhat** [comment]
Tagging is done.  See #360 for CHANGELOG :smiley_cat: :metal:.

File: cli-migrate-slice-types [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
```
Comment (jszwedko) [inline cli-migrate-slice-types]: :clap: this is above, beyond, and awesome.

File: flag.go [hunk 2]
```diff
@@ -10,6 +11,10 @@ import (
 	"time"
 )
 
+var (
+	slPfx = fmt.Sprintf("sl:::%d:::", time.Now().UTC().UnixNano())
```
Comment (jszwedko) [inline flag.go]: Why is this prefix needed?

Comment (meatballhat) [inline flag.go]: This is me being paranoid.  I'm trying to minimize the risk that we'd deserialize something coming out of the stdlib `flag` layer that's not "ours" to deserialize.  My concern isn't so much about safety, but rather aiming for clarity, avoiding unexpected bugs, making it super easy to spot when a string value from the cli layer isn't being handled correctly.

Comment (jszwedko) [inline flag.go]: Gotcha, that makes sense to me, do you mind adding a comment to that effect for _enhanced_ clarity?

Comment (jszwedko) [inline flag.go]: Otherwise :shipit:

**Comment by jszwedko** [comment]
@meatballhat what do you think about targeting a `2.0` branch with this so that we can continue to iterate against master while we prepare the 2.0 release? Alternatively, we could have a `1.x` branch that we backport to until 2.0 is officially released.

**Comment by meatballhat** [comment]
I'm a fan of a v2 branch, assuming the majority of humans who use cli are not using version pinning.

**Comment by meatballhat** [comment]
Superseded by #392

**Pull request closed without merging**